notification/slack/actions.go

```diff
@@ -0,0 +1,191 @@
+// https://api.slack.com/authentication/verifying-requests-from-slack
+func validRequest(w http.ResponseWriter, req *http.Request) bool {
+	if req.Method != "POST" {
+		http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+		return false
+	}
+
+	ts := req.Header.Get("X-Slack-Request-Timestamp")
+
+	// ignore request if more than 5 minutes from local time
+	_ts, err := strconv.ParseInt(ts, 10, 64)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusUnprocessableEntity), http.StatusUnprocessableEntity)
+		return false
+	}
+	if abs(time.Now().Unix()-_ts) > 60*5 {
+		return false
+	}
+
+	defer req.Body.Close()
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusUnprocessableEntity), http.StatusUnprocessableEntity)
+		return false
+	}
+	req.Body.Close()
+	req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+	secret := config.FromContext(req.Context()).Slack.SigningSecret
+	h := hmac.New(sha256.New, []byte(secret))
+	fmt.Fprintf(h, "v0:%s:%s", ts, body)
+	calculatedSignature := "v0=" + hex.EncodeToString(h.Sum(nil))
+	signature := []byte(req.Header.Get("X-Slack-Signature"))
+
+	return hmac.Equal(signature, []byte(calculatedSignature))
+}
+
+// ServeActionCallback processes POST requests from Slack. A callback ID is provided
+// to determine which action to take.
+func (h *Handler) ServeActionCallback(w http.ResponseWriter, req *http.Request) {
+	if !validRequest(w, req) {
+		return
+	}
+
+	var payload slack.InteractionCallback
+	err := json.Unmarshal([]byte(req.FormValue("payload")), &payload)
+	if err != nil {
+		panic(err)
+	}
+
+	// don't process url buttons
+	for _, action := range payload.ActionCallback.BlockActions {
+		if action.ActionID == "openLink" || action.ActionID == "auth" {
+			return
+		}
+	}
+
+	intrnlErr := func() { http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError) }
+	clientErr := func() { http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest) }
+	process := func(ctx context.Context) {
+		cfg := config.FromContext(ctx)
+		var api = slack.New(cfg.Slack.AccessToken)
+
+		// check if user valid, if ID does not exist return ephemeral to auth with GoAlert
+		_, err := h.c.UserStore.FindOneBySlackUserID(ctx, payload.User.ID)
+		if err != nil {
+			uri := cfg.General.PublicURL + "/api/v2/slack/auth"
+			msg := userAuthMessageOption(cfg.Slack.ClientID, uri)
+			_, err := api.PostEphemeral(payload.Channel.ID, payload.User.ID, msg)
+			if err != nil {
+				clientErr()
+				return
+			}
+			return
+		}
+
+		// actions may come in batches, range over
+		for _, action := range payload.ActionCallback.BlockActions {
+			alertID, err := strconv.Atoi(action.Value)
+			if err != nil {
+				intrnlErr()
+				return
+			}
+
+			// add source info to ctx to write to alert log
+			ncID, _, err := h.c.AlertLogStore.FindByValue(ctx, nil, payload.Channel.ID)
```

Comment:
Right now this only includes channel, but yeah since this is sourced from a user's action within Slack I think we will need both for the alert log

Reply:
I'm wondering if, since we will need both the User's name and the Channel's name, if we'll need to create a new source type, like "SourceTypeNotificationChannelUser" with two ID fields to use instead of the normal 1.

Reply:
Per internal discussion, moving forward with the user's name and leaving out the channel for MVP:
"Acknowledged by Nathaniel Cook (Slack)"

This is acceptable because other log events do not show phone numbers or emails, i.e. "Acknowledged by Nathaniel Cook (SMS)"